cpp/oneapi/dal/detail/serialization.hpp

```diff
@@ -0,0 +1,467 @@
+    }
+    return *serializable;
+}
+
+template <typename T>
+inline serializable_iface& get_serializable(const std::shared_ptr<T>& object) {
+    return get_serializable(object.get());
+}
+
+template <typename T>
+inline void serialize_polymorphic(T* serializable_object, output_archive& archive) {
+    const bool is_not_nullptr = (serializable_object != nullptr);
+    archive(is_not_nullptr);
+
+    if (is_not_nullptr) {
+        auto& serializable = get_serializable(serializable_object);
+        archive(serializable.get_serialization_id());
+        serializable.serialize(archive);
+    }
+}
+
+template <typename T>
+inline void serialize_polymorphic_shared(const std::shared_ptr<T>& serializable_object,
+                                         output_archive& archive) {
+    serialize_polymorphic(serializable_object.get(), archive);
+}
+
+template <typename T>
+inline T* deserialize_polymorphic(input_archive& archive,
+                                  const std::initializer_list<std::uint64_t>& allowed_ids = {}) {
+    const bool is_not_nullptr = archive.pop<bool>();
+    if (!is_not_nullptr) {
+        return nullptr;
+    }
+
+    const auto serialization_id = archive.pop<std::uint64_t>();
+    if (allowed_ids.size() > 0) {
+        const auto it = std::find(allowed_ids.begin(), allowed_ids.end(), serialization_id);
+        if (it == allowed_ids.end()) {
+            throw invalid_argument{ error_messages::archive_content_does_not_match_type() };
+        }
+    }
+
+    T* object_ptr = serializable_registry::instance().make<T>(serialization_id);
+    get_serializable(object_ptr).deserialize(archive);
+    return object_ptr;
+}
+
+template <typename T>
+inline std::shared_ptr<T> deserialize_polymorphic_shared(
+    input_archive& archive,
+    const std::initializer_list<std::uint64_t>& allowed_ids = {}) {
+    return std::shared_ptr<T>{ deserialize_polymorphic<T>(archive, allowed_ids) };
+}
+
+template <typename T>
+inline void deserialize_polymorphic_shared(
+    std::shared_ptr<T>& serializable_object,
+    input_archive& archive,
+    const std::initializer_list<std::uint64_t>& allowed_ids = {}) {
+    serializable_object.reset(deserialize_polymorphic<T>(archive, allowed_ids));
+}
+
+template <typename T, typename OutputArchive>
+inline void serialize(const T& value, OutputArchive& archive) {
+    output_archive internal_archive{ archive };
+    internal_archive.prologue();
+    internal_archive(value);
+    internal_archive.epilogue();
+}
+
+template <typename T, typename InputArchive>
+inline void deserialize(T& value, InputArchive& archive) {
+    input_archive internal_archive{ archive };
+    internal_archive.prologue();
+    internal_archive(value);
+    internal_archive.epilogue();
+}
+
+class binary_output_archive : public base {
+public:
+    void prologue() {}
+    void epilogue() {}
+    void operator()(const void* data, data_type dtype, std::int64_t count = 1) {}
+};
+
+class binary_input_archive : public base {
```

Comment:
What are these entities responsible for?

Reply:
Now it's implemented